File: src/core/entity-builder.ts
import { tool } from "langchain";
import { z } from "zod";
import type { GraphmindStore } from "../db/client.js";

/**
 * Creates a `create_entity` tool that lets agents create arbitrary entities
 * in the context graph — building their "brain map" over time.
 *
 * Agents should use `inspect_schema` first to understand what already exists,
 * then create new entity types only when genuinely new concepts are discovered.
 *
 * Examples of entities agents might create:
 * - A coding agent: CodeFile, Function, APIEndpoint, Dependency, BugReport
 * - A legal agent: Contract, Clause, Regulation, CaseReference
 * - A medical agent: Condition, Medication, Symptom, LabResult
 */
export function createEntityTool(store: GraphmindStore) {
  return tool(
    async (input) => {
      const now = new Date().toISOString();
      const agentName = store.getAgentName();
      const { label, reason, properties: explicitProps, ...extraKeys } = input as any;

      // LLMs often pass properties as top-level args instead of inside `properties`.
      // Collect any extra keys (path, name, decision, etc.) into the properties object.
      const props: Record<string, string | number | boolean> = {
        ...(explicitProps ?? {}),
      };
      for (const [key, val] of Object.entries(extraKeys)) {
        if (typeof val === "string" || typeof val === "number" || typeof val === "boolean") {
          props[key] = val;
        }
      }

      try {
        const allProps = { ...props };
        if (reason) allProps._reason = reason;

        const entityId = await store.createEntity({
          label,
          properties: allProps,
          createdBy: agentName,
          createdAt: now,
        });

        return `Entity created: ${label} (id: ${entityId}). Properties: ${JSON.stringify(props)}. This entity is now part of your context graph and can be referenced in relationships.`;
      } catch (err) {
        return `Failed to create entity: ${(err as Error).message}. Check that the label is valid (PascalCase, no special characters).`;
      }
    },
    {
      name: "create_entity",
      description:
        "Create a new entity (node) in the context graph to map your understanding of the domain. " +
        "Use this to record domain concepts, objects, or artifacts you discover while working. " +
        "Always use `inspect_schema` first to check if a similar entity type already exists. " +
        "Labels should be PascalCase (e.g., CodeFile, APIEndpoint, Contract). " +
        "Pass entity properties directly as named arguments (e.g., path, name, decision).",
      schema: z.object({
        label: z
          .string()
          .describe("PascalCase node label (e.g., 'CodeFile', 'APIEndpoint', 'Contract', 'Symptom')"),
        reason: z
          .string()
          .optional()
          .describe("Why you are creating this entity — what understanding does it capture?"),
      }).passthrough(),
    }
  );
}

/**
 * Creates a `create_relationship` tool that lets agents connect entities
 * in the context graph with meaningful relationships.
 *
 * Examples:
 * - (CodeFile)-[:IMPORTS]->(CodeFile)
 * - (Function)-[:HANDLES]->(ErrorType)
 * - (Contract)-[:GOVERNED_BY]->(Regulation)
 * - (Symptom)-[:INDICATES]->(Condition)
 */
export function createRelationshipTool(store: GraphmindStore) {
  return tool(
    async ({ source_id, target_id, relationship_type, properties, reason }) => {
      const now = new Date().toISOString();
      const agentName = store.getAgentName();

      try {
        const relProps: Record<string, string | number | boolean> = { ...properties };
        if (reason) relProps._reason = reason;

        await store.createRelationship({
          sourceId: source_id,
          targetId: target_id,
          type: relationship_type,
          properties: relProps,
          createdBy: agentName,
          createdAt: now,
        });

        return `Relationship created: (${source_id})-[:${relationship_type}]->(${target_id}). This connection is now part of your context graph.`;
      } catch (err) {
        return `Failed to create relationship: ${(err as Error).message}. Verify that both source and target IDs exist.`;
      }
    },
    {
      name: "create_relationship",
      description:
        "Create a relationship between two existing entities in the context graph. " +
        "Use this to map connections, dependencies, and associations between domain concepts. " +
        "Relationship types should be UPPER_SNAKE_CASE (e.g., IMPORTS, DEPENDS_ON, TREATS, GOVERNED_BY).",
      schema: z.object({
        source_id: z.string().describe("The node ID of the source entity"),
        target_id: z.string().describe("The node ID of the target entity"),
        relationship_type: z
          .string()
          .describe("UPPER_SNAKE_CASE relationship type (e.g., 'IMPORTS', 'DEPENDS_ON', 'TREATS')"),
        properties: z
          .record(z.string(), z.union([z.string(), z.number(), z.boolean()]))
          .optional()
          .describe("Optional properties on the relationship edge"),
        reason: z
          .string()
          .optional()
          .describe("Why this relationship exists — what connection does it capture?"),
      }),
    }
  );
}

/**
 * Creates a `find_entities` tool for searching existing entities by label and properties.
 */
export function createFindEntitiesTool(store: GraphmindStore) {
  return tool(
    async ({ label, filter }) => {
      try {
        const entities = await store.findEntities(label, filter as Record<string, string | number | boolean> | undefined);

        if (entities.length === 0) {
          return `No ${label} entities found${filter ? ` matching ${JSON.stringify(filter)}` : ""}. Use create_entity to add one.`;
        }

        const items = entities.map((e) => {
          const props = Object.entries(e.properties)
            .filter(([k]) => !k.startsWith("_"))
            .map(([k, v]) => `${k}: ${v}`)
            .join(", ");
          return `- id: ${e.id} | ${props}`;
        });

        return `Found ${entities.length} ${label} entit${entities.length === 1 ? "y" : "ies"}:\n${items.join("\n")}`;
      } catch (err) {
        return `Search failed: ${(err as Error).message}`;
      }
    },
    {
      name: "find_entities",
      description:
        "Search for existing entities in the context graph by label and optional property filter. " +
        "Use this to find entities before creating duplicates or to look up IDs for creating relationships.",
      schema: z.object({
        label: z.string().describe("The entity label to search for (e.g., 'CodeFile', 'Contract')"),
        filter: z
          .record(z.string(), z.union([z.string(), z.number(), z.boolean()]))
          .optional()
          .describe("Optional property filter (e.g., {name: 'auth.ts'})"),
      }),
    }
  );
}

File: python-sdk/src/graphmind_context_graphs/core/entity_builder.py
"""Dynamic entity builder tools for agent brain mapping."""

from __future__ import annotations

import json
from datetime import datetime, timezone

from langchain.tools import tool
from pydantic import BaseModel, Field, ConfigDict

from ..db.client import GraphmindStore
from ..types.data_model import GraphEntity, GraphRelationship


class CreateEntityInput(BaseModel):
    """Input for create_entity. Extra fields become entity properties."""
    model_config = ConfigDict(extra="allow")

    label: str = Field(description="PascalCase node label (e.g., 'CodeFile', 'APIEndpoint', 'Contract')")
    reason: str | None = Field(default=None, description="Why you are creating this entity")


def create_entity_tool(store: GraphmindStore):
    @tool("create_entity", args_schema=CreateEntityInput)
    def create_entity(**kwargs) -> str:
        """Create a new entity in the context graph to map your understanding of the domain.
        Pass entity properties directly as named arguments (e.g., path, name, decision)."""
        label = kwargs.pop("label")
        reason = kwargs.pop("reason", None)
        now = datetime.now(timezone.utc).isoformat()

        # All remaining kwargs become entity properties
        props: dict[str, str | int | float | bool] = {}
        for k, v in kwargs.items():
            if isinstance(v, (str, int, float, bool)):
                props[k] = v

        if reason:
            props["_reason"] = reason

        try:
            entity_id = store.create_entity(GraphEntity(
                label=label,
                properties=props,
                created_by=store.agent_name,
                created_at=now,
            ))
            return f"Entity created: {label} (id: {entity_id}). Properties: {json.dumps(props)}."
        except Exception as e:
            return f"Failed to create entity: {e}"

    return create_entity


class CreateRelationshipInput(BaseModel):
    source_id: str = Field(description="The node ID of the source entity")
    target_id: str = Field(description="The node ID of the target entity")
    relationship_type: str = Field(description="UPPER_SNAKE_CASE relationship type (e.g., 'IMPORTS', 'DEPENDS_ON')")
    reason: str | None = Field(default=None, description="Why this relationship exists")


def create_relationship_tool(store: GraphmindStore):
    @tool("create_relationship", args_schema=CreateRelationshipInput)
    def create_relationship(source_id: str, target_id: str, relationship_type: str, reason: str | None = None) -> str:
        """Create a relationship between two existing entities in the context graph."""
        now = datetime.now(timezone.utc).isoformat()
        props: dict[str, str | int | float | bool] = {}
        if reason:
            props["_reason"] = reason

        try:
            store.create_relationship(GraphRelationship(
                source_id=source_id,
                target_id=target_id,
                type=relationship_type,
                properties=props,
                created_by=store.agent_name,
                created_at=now,
            ))
            return f"Relationship created: ({source_id})-[:{relationship_type}]->({target_id})."
        except Exception as e:
            return f"Failed to create relationship: {e}"

    return create_relationship


class FindEntitiesInput(BaseModel):
    label: str = Field(description="The entity label to search for (e.g., 'CodeFile')")
    filter: dict[str, str | int | float | bool] | None = Field(default=None, description="Optional property filter")


def create_find_entities_tool(store: GraphmindStore):
    @tool("find_entities", args_schema=FindEntitiesInput)
    def find_entities(label: str, filter: dict | None = None) -> str:
        """Search for existing entities in the context graph by label and optional property filter."""
        try:
            entities = store.find_entities(label, filter)
            if not entities:
                return f"No {label} entities found. Use create_entity to add one."

            items = []
            for e in entities:
                visible_props = {k: v for k, v in e.properties.items() if not str(k).startswith("_")}
                prop_str = ", ".join(f"{k}: {v}" for k, v in visible_props.items())
                items.append(f"- id: {e.id} | {prop_str}")

            return f"Found {len(entities)} {label} entit{'y' if len(entities) == 1 else 'ies'}:\n" + "\n".join(items)
        except Exception as e:
            return f"Search failed: {e}"

    return find_entities
